### ISSUE TITLE
Maybe GC optimization in Unity?

### ISSUE THREAD

**Issue by BoysheO** [Cysharp/R3#72]
Happy to see R3 created.I have used UniRx years.UniRx use Gc a lot.How about reduce the memory usage or make some thing pooled?
I have a simple solution but not been test and validated,it still a toy.see [BoysheO.Extensions](https://github.com/BoysheO/BoysheO.Extensions/tree/main)/[src](https://github.com/BoysheO/BoysheO.Extensions/tree/main/src)/[UnityReactive.Core](https://github.com/BoysheO/BoysheO.Extensions/tree/main/src/UnityReactive.Core)

What's it design?
 "event object whitch is oop"

1. The IObserver.OnError is not very useful in practice. In situations where handling errors is a common requirement, such as dealing with network-related issues, using UniTask is more beneficial and easier to debug. IObserver appears more as an alternative to C#'s events, providing abstract capabilities that events lack. Personally, my implementation leans towards removing the OnError interface. If an event source wants to push an error notification, it is advisable to consider wrapping the error within an element, for instance, returning something like OneOf{Exception, Element}.

2.The IObservable.Subscribe method currently returns an IDisposable object, which can be prone to consuming garbage collection resources. I believe designing the interface as IObservable.Subscribe + IObservable.Unsubscribe is appropriate. However, a challenge arises if Subscribe is called multiple times, as IObservable.Unsubscribe might struggle to distinguish which listener to operate on. Perhaps changing the IDisposable returned by IObservable.Subscribe to return a value object could be a suitable solution. This way, it would only require maintaining an ID table internally within the Observable object.Certainly, my approach is more aggressive. I use a Manager class to maintain all the ID tables, instead of maintaining them within the Observable.

3.An event dispatched to be published elsewhere should have tracking information. My approach is to simply add a trackingMessage to the Observable. I believe this would be immensely helpful for debugging. Otherwise, when an IObserver throws an error, we can only see the stack trace printed from the scheduler.

4.After developing my UnityReactive library to address the aforementioned issues, I've encountered a dilemma. Using Linq inevitably creates a class of objects like "middlewares," leading to increased garbage collection (GC) overhead. This poses a challenge for UnityReactive managed by UnityReactiveManager. Perhaps I should implement a parsing language similar to SQL or create a separate Manager for each operator to manage their respective states. I haven't decided which approach to take yet; it seems both have their drawbacks and may not be ideal.

Although it might sound repetitive, I want to emphasize the GC issue. In Unity, this is crucial. Striking the optimal balance between GC and usability is a must.

**Comment by neuecc** [comment]
Thank you for your interesting implementation and insights.
I generally agree, and in R3 we have gone back to the concept of enhancing events, removing OnError and so on.

We are not considering approaches like pooling at the moment.
Since the chains built with Observable are long-lived, we believe that the allocations made at construction time can be disregarded.
Instead, for things like one-off asynchronous methods that are short-lived and likely to be repeated frequently, we recommend using UniTask.
In R3, we are focusing on integration with async/await.

However, additional allocations within OnNext must be avoided.
In UniRx, there were several instances that caused allocations every OnNext.
In R3, it's zero (probably!)
IScheduler was also a source of allocations, so it has been removed.

Additionally, we have minimized the allocation at the time of Subscribe itself.
We have made sure that the Observer itself becomes IDisposable, and there are hardly any additional IDisposable creations.
